README.md
uTask is a small library that makes it easier to write state machines in embedded CPUs.

How to use uTask library:
1. Configure uTask using utask.h
2. Change your software to make use of uTask.
3. Link your software with utask!

CHANGELOG.md
Version 0.3.0 [2010-8 - not released]:
- Changed several things since 0.2.2 version
	
Version 0.3.1 [2010-10-15]:
- Prepared for release
- Added examples
	
Version 0.3.2 [2011-02-28]:
- [**fix**] Fixed bug in utask_add() with UTASK_SIMPLE_ADD defined. Only the first task would run.
- [**new**] Included typedefs for easy change of utask's member size. Code is prepared for 8 and 16-bit uCs.
- [**new architecture**] uTasks have been confirmed to work on Atmel ATmega8 device.

Version 0.3.3 [2011-06-04]:
- [**fix**] Fixed bug in utask_schedule(), task could be woken prematurely,
- [**thanx**] for Mateusz Kondej for finding utask_schedule() bug,
- [**new**] Added function to check if CPU can be put to sleep, user needs to implement it's own function to put CPU to sleep.
	

utask.c
/*
	 Copyright (c) 2010 Bartosz Bielawski, Marcin Zapolski
	 Copyright (c) 2011 Bartosz Bielawski
	This software is licensed under MIT License, see LICENSE file.
*/


#include <stdint.h>
#include "utask.h"


/**
 @file utask.c
 @authors B.Bielawski - InventLab, Marcin Zapolski - InventLab
 @date 2010.02.28
 @version 0.3.3
*/


static utask_data_t utask_free_slots = UTASKS;
static utask_data_t utask_free_tid = 0;
static utask_data_t utask_last_used = 0;

#ifdef UTASK_TID
	static utask_data_t utask_last_tid = 0;
#endif

static utask_t 	tasks[UTASKS];

#ifdef UTASK_CHECK_MCU_SLEEP
	void (*utask_put_mcu_to_sleep)(utask_data_t d) = (void*)0;
#endif

void utask_init()
{
    int i;
    for (i=0;i<UTASKS;i++)
		tasks[i].state = UTS_NONE;
}

void utask_sleep(utask_t* t,utask_timer_t ticks)
{
    t->sleep = ticks;
    t->state = UTS_SLEEP;
    #ifdef UTASK_SEM
		t->sem = 0;
	#endif
}


#ifdef UTASK_NAME
utask_t*	utask_add_name(utask_fun fun, const char* name)
{
	utask_t* n = utask_add(fun);
	if (n)
		n->name = name;
	return n;
}
#endif

utask_t*	utask_add(utask_fun fun)
{
	utask_t* t = &tasks[utask_free_tid];

	if (!utask_free_slots)
		return 0;			//NULL

   	t->state = UTS_READY;
    t->istate = 0;
    t->sleep = 0;
	t->fun = fun;
	
	#ifdef UTASK_ARG
		t->arg = 0;
	#endif

	#ifdef UTASK_NAME
		t->name = "";
	#endif

	#ifdef UTASK_TID
		t->tid = utask_last_tid++;
	#endif

	#ifdef UTASK_SIMPLE_ADD
		utask_free_tid++;
		utask_free_slots--;
		utask_last_used++;
	#else

	if (utask_last_used < utask_free_tid)
		utask_last_used = utask_free_tid;		//nowy ostatni element

	int i;

    for (i=0;i<UTASKS;i++)			//wyznacz nowy pierwszy pusty element
		if (tasks[i].state == UTS_NONE)
		{
			utask_free_tid = i;
			break;
		}
	utask_free_slots--;
	#endif
	return t;
}


#ifdef UTASK_CHECK_MCU_SLEEP
utask_data_t utask_check_sleep()
{
	utask_t* t = tasks;
	utask_data_t i;
	utask_timer_t s = UTASK_SLEEP_MAX;
	for (i=0;i<=utask_last_used;++i,++t)	//iterate over used tasks
	{
		if (t->state == UTS_READY)
			return 0;
		if (t->state >= UTS_SLEEP)
			if (t->sleep < s)
				s = t->sleep;
	}	
	return s;
}
#endif

void utask_schedule()
{
	int i;
#ifdef UTASK_SEM
	register utask_sem_t  cond = 0;
#endif
	register utask_t* t = tasks;
	for (i=0;i<=utask_last_used;++i,++t)	//iterate over used tasks
	{
		switch (t->state)
		{
			case UTS_NONE:
				continue;
			case UTS_SUSP:
				continue;
				
			case UTS_READY:			//this one is ready for execution
				t->fun(t);
				break;

			case UTS_SLEEP:
				continue;

#ifdef UTASK_SEM
	#ifdef UTASK_SEM_CHANGE
			case  UTS_WAIT_E:		//check if semaphore's value is equal to the one we are waiting for
				cond = (*t->sem == t->sem_val);
				break;
			case  UTS_WAIT_NE:		//inverted condition
				cond = (*t->sem != t->sem_val);
				break;
	#else
			case UTS_WAIT_Z:		//check if semaphore's value is zero
				cond = !(*t->sem);
				break;
			case UTS_WAIT_NZ:		//check if semaphore's value is non-zero
				cond = (*t->sem);
				break;
	#endif
#endif
			//just in case
			default:			//this catches UTS_NONE, UTS_SUSP, UTS_SLEEP	
				break;
		}
	
		
#ifdef UTASK_SEM
		if (cond)
		{
			t->state = UTS_READY;
			t->fun(t);
		}
#endif	

	}
	
#ifdef UTASK_CHECK_MCU_SLEEP
	if (!utask_put_mcu_to_sleep)		//if no callback has been defined return
		return;		
	
	utask_timer_t s = utask_check_sleep();
	utask_put_mcu_to_sleep(s);
#endif
}


void utask_sleep_process()
{
	int i;
	for (i=0;i<=utask_last_used;i++)
	{
		if (tasks[i].state < UTS_SLEEP)			//nie zyje, do wykonania, zawieszony
			continue;

		if (tasks[i].sleep > 0)
			tasks[i].sleep--;
		else
			tasks[i].state = UTS_READY;
	}
}


#ifdef UTASK_EXIT
void utask_exit(utask_t* t)
{
	int i;
	t->state = UTS_NONE;
	utask_free_slots++;
	utask_free_tid = UTASKS;
	utask_last_used = 0;
	for (i=0;i<UTASKS;i++)
	{
		if (tasks[i].state == UTS_NONE)
		{
			if (utask_free_tid > i)
				utask_free_tid = i;
		}
		else
			utask_last_used = i;
	}
}
#endif

#ifdef UTASK_SEM 
	#ifdef UTASK_SEM_CHANGE

//czekanie na jakas wartosc semafora
void utask_wait_eq(utask_t* t, utask_sem_t* sem, utask_sem_t sem_val, utask_timer_t timeout)
{
	t->sem = sem;
	t->sleep = timeout;
	t->state = UTS_WAIT_E;
	t->sem_val = sem_val;
}

//czekanie na zmiane z jakiejs wartosci semafora
void utask_wait_neq(utask_t* t, utask_sem_t* sem, utask_sem_t sem_val, utask_timer_t timeout)
{
	t->sem = sem;
	t->sleep = timeout;
	t->state = UTS_WAIT_NE;
	t->sem_val = sem_val;
}

//owrapowanie dla zera
#define utask_wait_nzero(utask, sem, timeout)	utask_wait_neq(utask, sem, 0, timeout)
#define utask_wait_zero(utask, sem, timeout)	utask_wait_eq(utask, sem, 0, timeout)
	#else
	
//to samo, czekanie dla zera i niezera
void utask_wait_nzero(utask_t* t, utask_sem_t* sem, utask_timer_t timeout)
{
	t->sem = sem;				//semafor
	t->sleep = timeout;
	t->state = UTS_WAIT_NZ;
}

void utask_wait_zero(utask_t* t, utask_sem_t* sem, utask_timer_t timeout)
{
	t->sem = sem;				//semafor
	t->sleep = timeout;
	t->state = UTS_WAIT_Z;
}	
	
	#endif
#endif

#ifdef UTASK_SUSPEND_RESUME
void	utask_suspend(utask_t* t)
{
	t->state = UTS_SUSP;
}

void	utask_resume(utask_t* t)
{
	t->state = UTS_READY;
	t->sleep = 0;
}
#endif

#ifdef UTASK_STATS
utask_data_t	utask_get_free_slots(void)
{
	return utask_free_slots;
}

utask_data_t    utask_get_task_cnt(utask_fun fun)
{
    int i;
    int cnt = 0;
	for (i=0;i<utask_last_used;i++)
		if (tasks[i].fun == fun)
			cnt++;
	return cnt;
}
#endif


#ifdef UTASK_TID
utask_t* utask_get_by_tid(utask_data_t tid)
{
	int i;
	for (i=0;i<utask_last_used;i++)
		if (tasks[i].tid == tid)
			return tasks+i;
	return 0;
}
#endif


utest.c
#include <stdio.h>
#include <pthread.h>
#include <unistd.h>
#include "utask.h"
#include <signal.h>

pthread_t timer_thread;

volatile int run = 1;

#define TIPS 10

#define D1S		TIPS
#define D3S		(3*D1S)
#define D500MS	(D1S/2)
#define D100MS  (D1S/10)

void* timer_thread_fun(void* arg)
{
	while (run)
	{
		utask_sleep_process();
		usleep(1e6 / TIPS);
	}
	return;
}


int semaphore = 0;

void task_print(utask_t* t)
{
	switch (t->istate)
	{
		case 0:
			t->istate = 1;
			utask_wait_eq(t,&semaphore,0,D3S);
			break;
		
		case 1:
			if (t->sleep == 0)
			{
				t->istate = 0;
				return;
			}
		
			semaphore = 1;
			
			printf("%d\n",t->arg);
			utask_sleep(t,D1S);
			t->istate = 2;
			break;
		case 2:
		
			semaphore = 0;
			t->istate = 0;
			utask_sleep(t,D100MS);
			break;
	}
}

void task_timedout(utask_t* t)
{
	switch (t->istate)
	{
		case 0:
			utask_wait_eq(t,&semaphore,0,D500MS);
			t->istate = 1;
			break;
			
		case 1:
			if (t->sleep == 0)
			{
				printf("T\n");
				t->istate = 0;
			}
			else
			{
				t->istate = 2;
				semaphore = 1;
				utask_sleep(t,D500MS);
			}
				

			break;
		case 2:
			printf("OK\n");
			semaphore = 0;
			t->istate = 0;
			utask_sleep(t,D1S);
			break;
	}
}

void signal_handler(int sig)
{
	fprintf(stderr,"Got signal!\n");
	run = 0;
}

void mcu_sleep(utask_timer_t s)
{
	if (s == 0)
		return;
	printf("%dT\n",s);
	usleep((1e6 * s / TIPS) / 2 + 1000);
}

int main()
{
	utask_init();

	signal(SIGINT,signal_handler);
	signal(SIGQUIT, signal_handler);
	signal(SIGTERM, signal_handler);

	pthread_create(&timer_thread,NULL,timer_thread_fun,NULL);
	
	utask_t* t;
	t = utask_add(task_print);
	t->arg = 1;
	t = utask_add(task_print);
	t->arg = 2;
	t = utask_add(task_print);
	t->arg = 3;
	
	utask_add(task_timedout);
	
	utask_put_mcu_to_sleep = mcu_sleep;
	
	while (run)
	{
		utask_schedule();
	}
	
	pthread_join(timer_thread,NULL);
	return 0;
}



examples/not_ready/e6.c
#include "utask.h"
#include "leds.h"
#include "timer.h"
#include "uart.h"

#define SECOND1000

//called 1000 times a second
void timer_isr()
{
	utask_sleep_process();
}


#define WT_INIT0
#define WT_WORKING1
#define WT_READY2

#define LONG_LONG_TIME 0xFFFFFFFF

void worker_task(utask_t * t)
{
	switch (t->istate)
	{
		case 0:
//working task is alive!
			t->arg = WT_WORKING;
			t->istate = 1;
			break;

		case 1:
/*
Do the heavy work here.
It can take some time.
.
.
.
*/
			break;

/*
.
.
.
/*

case N-1:
/*
...
*/
			break;

		case N:
//set ready flag for parent process
//and wait to be killed
			t->data[0] = WT_READY;
			utask_sleep(t, LONG_LONG_TIME);
			break;
	}
}


void main_task(utask_t * t)
{
	static utask_t *worker = NULL;
	switch (t->istate)
	{
		case 0:
			uart_puts("Hello :)\n");
			t->istate = 1;
			break;

		case 1:
//create new task
			utask * worker = utask_add(worker_task);
//assign a new state
			worker->arg = WT_INIT;
//wait for worker state to be set to WT_READY
			t->istate = 2;
			utask_wait_eq(t, &worker->arg, WT_READY, 5 * SECOND) break;

		case 2:
			if (t->sleep)
			{
/*
collect results here
*/
			}
//kill worker task
			utask_exit(worker);
//go to the first state again
			t->istate = 1;
			utask_sleep(t, 5 * SECOND);
			break;
	}
}


int main()
{
	timer_init();
	utask_init();

	utask_add(main_task);

	while (1)
	{
		utask_schedule();
	}
	return 0;
}


examples/utask_example_2.c
#include "utask.h"
#include "leds.h"
#include "timer.h"

void timer_isr ()			//called 1000 times a second
{
  utask_sleep_process();
}

void task_delayed_run(utask * t)
{
	turn_led_on(t->arg);		//turn led on
	utask_exit(t);		//end task
}

int main ()
{
	timer_init();
	utask_init();
	
	utask* new_task;
	
	new_task = utask_add (task_delayed_run);
	utask_sleep(new_task,5000);		//delay execution, it will run in 5 seconds
	new_task->arg = RLED;			//turn on red led
	
	new_task = utask_add (task_delayed_run);
	utask_sleep(new_task,10000);		//delay execution, it will run in 10 seconds
	new_task->arg = GLED;			//turn on green led
	
	new_task = utask_add (task_delayed_run);
	utask_sleep(new_task,15000);		//delay execution, it will run in 15 seconds
	new_task->arg = BLED;			//turn on blue led
	
	while (1)
	{
		utask_schedule();
	}
	return 0;
}


examples/utask_example_3.c
#include "utask.h"
#include "leds.h"
#include "timer.h"
#include "uart.h"

void timer_isr()			//called 1000 times a second
{
  utask_sleep_process();
}

int uart_busy = 0;

void task_send_msg(utask* t)
{
	switch (t->istate)
	{
		case 0:
			utask_wait_zero(t,&uart_busy,1000);	//wait for uart to be free, timeout 1000ms
			t->istate = 1;
			break;
			
		case 1:
			if (t->sleep != 0)		//task has been awakened, uart_busy == 0
			{
				uart_busy = 1;	//claim resource
				t->istate = 2;
			}
			else
				t->istate = 0;		//go to waiting loop again
			break
			
		case 2:
			//do yer stuff with uart
			/*
			.
			.
			.
			*/
			break;
			
		case N-1:
			//and here
			/*
			.
			.
			.
			*/
			break;
		
		case N:
			uart_busy =  0;		//free resource
			utask_sleep(t,5000);
			t->istate = 0;		//do that over and over again
			break;
	}
}

const char* strings[] = 
{
	"Oh hello!",
	"Howdy!",
	"Fhtagn?"
};


int main ()
{
	timer_init();
	uart_init();
	utask_init();
	
	utask* t;
	
	t = utask_add(task_send_msg);		//add task
	t->arg = 0;		//use 1st string
	
	t = utask_add(task_send_msg);		//and another one
	t->arg = 1;		//use 2nd string
	
	t = utask_add(task_send_msg);		//and another one
	t->arg = 2;		//use 3rd string
	
	while (1)
	{
		utask_schedule();
	}
	return 0;
}


examples/utask_example_4.c
#include "utask.h"
#include "leds.h"
#include "timer.h"
#include "uart.h"

void timer_isr()			//called 1000 times a second
{
  utask_sleep_process();
}

int btn_press_counter = 0;

void button_isr()
{
	btn_press_counter++;
}

void task_do_something(utask* t)
{
	//do yer stuff here
}

void task_spawn_on_button_press(utask* t)
{
	switch (t->istate)
	{
		case 0:
			utask_wait_nzero(t,&btn_press_counter,1000000);		//wait looooong time
			t->istate = 1;
			break;
			
		case 1:
			if (t->sleep != 0)	//the task has been awakened -> btn_press_counter != 0
			{
				utask_add(task_do_something);		//spawn new task
				btn_press_counter--;
			}
			t->istate = 0;
			break;
	}
}


int main ()
{
	timer_init();
	uart_init();
	utask_init();

	utask_add(task_spawn_on_button_press);
	
	while (1)
	{
		utask_schedule();
	}
	return 0;
}


utask.h
/*
	 Copyright (c) 2010 Bartosz Bielawski, Marcin Zapolski
	 Copyright (c) 2011 Bartosz Bielawski
	This software is licensed under MIT License, see LICENSE file.
*/


#ifndef UTASK_H
#define UTASK_H

/**
 @file utask.h
 @authors B.Bielawski - InventLab, Marcin Zapolski - InventLab
 @date 2010.02.28
 @mainpage
 @version 0.3.3

 No long introductions. Just short Q&A list. Feel free to ask questions. The newest version of uTask can be downloaded at https://sourceforge.net/projects/utask/.

 @b Q&A
 
	@b Q: What is uTask?
	
	@b A: uTask is a simple library providing simple coroutines (http://en.wikipedia.org/wiki/Coroutine).
 
 	@b Q: Why was uTask created?
 	
 	@b A: The library's main goal is to make writing firmware for 32bit uCs easier by providing simple framework for running multiple task in non-colliding manner. 
 	
 	@b Q: Can I use uTask on other devices? @b [new]
 	
 	@b A: I have used uTask on ATmega8 uC, for better performance tune the structure a little, reduce members' sizes.
 	
 	@b Q: How does it work?
 	
 	@b A: uTask manages user's FSMs (http://en.wikipedia.org/wiki/Finite-state_machine) implemented as functions. uTask provides functions to put tasks to sleep, suspend them or make them wait for specific resource to be free. It also can spawn new tasks at runtime and end them if they are no longer needed.
 	
 	@b Q: How big is uTask?
 	
 	@b A: Library's footprint can be reduced to about 600 B of flash and 200 B of RAM, depending on selected features and maximal number of tasks that can be run concurrently. Full blown version occupies about 1.5 kB of flash and about 40B of RAM per task (again, depending on user choices).
 	
 	@b Q: Was uTask tested?
 	
 	@b A: Yes, it was tested for a few months now, it is used in production equipment.
 	
 	@b Q: Is uTask free software?
 	
 	@b A: Yes, uTask is a free software, it is licensed under MIT License (see LICENSE file).
 	
 	@b Q: Why is it free?
 	
 	@b A: It is free because I like it that way and my boss agreed to release it to the public. This is just a small tool of mine, it's not a commercial product. Enjoy!
 	
 	@b Q: Who wrote uTask?
 	
 	@b Q: uTask was written by Bartosz Bielawski with help from Marcin Zapolski. We are both Embedded System Engineers working at InventLab (http://inventlab.eu) company. This software was created in our worktime.
 	
 
*/

//------------------------ USER CONFIG STARTS HERE ----------------------------------

#include <stdint.h>

///defines utask version string
#define UTASK_VER_STR	"0.3.3"
///major version
#define UTASK_VER_MAJOR 0
///minor version
#define UTASK_VER_MINOR 3
///revision
#define UTASK_VER_REV   3

///enables simplified utask_add (for fixed number of tasks)
#define UTASK_SIMPLE_ADD

///enables utask_exit for ending tasks
#define UTASK_EXIT

///enables task finding and counting
#define UTASK_STATS

///enables utask_suspend and utask_resume functions
#define UTASK_SUSPEND_RESUME

///emables task names
#define UTASK_NAME

///enables task indentifiers (tids)
#define UTASK_TID

///enables arg field in utask type
#define UTASK_ARG

///enables semaphores
#define UTASK_SEM

///enables waiting for specific semaphore values (without only for 0 and non zero)
#define UTASK_SEM_CHANGE

///enables data field in utask type (for free use)
#define UTASK_DATA		1

///defines how many tasks can be run in one time
#define UTASKS		10

///enables structure packing
#define UTASK_PACKED_STRUCT

///enables putting device to sleep
#define UTASK_CHECK_MCU_SLEEP

///This typedef defines size of utask_t->sleep timer member
typedef uint32_t utask_timer_t;
///This typedef defines size of utask_t->state and utask_t->istate member
typedef uint32_t utask_state_t;
///This typedef defines size of utask_t->sem member
typedef uint32_t utask_sem_t;
///This typedef defines size of utask_t->data member
typedef uint32_t utask_data_t;

#define UTASK_SLEEP_MAX 0xFFFFFFFF

/*
//suggested sizes for 16-bit uCs (MSP430, PIC24...)
typedef uint32_t utask_timer_t;
typedef uint16_t utask_state_t;
typedef uint16_t utask_istate_t;
typedef uint16_t utask_sem_t;
typedef uint16_t utask_data_t;	
#define UTASK_SLEEP_MAX 0xFFFFFFFF

//suggested sizes for 8-bit uCs (ATmega, AT90, PIC10/12/14/16/18)
typedef uint16_t utask_timer_t;
typedef uint8_t  utask_state_t;
typedef uint8_t  utask_istate_t;
typedef uint8_t  utask_sem_t;
typedef uint8_t  utask_data_t;
#define UTASK_SLEEP_MAX 0xFFFF
*/


//------------------------ END OF USER CONFIG ----------------------------------


#ifdef UTASK_SEM_CHANGE
	#ifndef UTASK_SEM
		#define UTASK_SEM
	#endif
#endif

///unsused
#define	UTS_NONE	0
///suspended
#define	UTS_SUSP	1
///ready for execution
#define UTS_READY	2
///asleep
#define UTS_SLEEP	3

#ifdef UTASK_SEM
#ifndef UTASK_SEM_CHANGE
	///waiting for semaphore == 0
	#define UTS_WAIT_Z	4
	///waiting for semaphore != 0
	#define UTS_WAIT_NZ	5
#else
	///waiting for semaphore == val (val set in function)
	#define UTS_WAIT_E 4
	///waiting for semaphore != val (val set in function)
	#define UTS_WAIT_NE 5
#endif
#endif

#ifdef UTASK_PACKED_STRUCT
	#pragma pack(1)
#endif

#ifdef UTASK_CHECK_MCU_SLEEP
	extern void (*utask_put_mcu_to_sleep)(utask_data_t d);
#endif


/**
 * @struct utask
 * @brief Structure describing single task
 *
 * This structure describes single task. Pointer to it is passed to many utask_* functions. Structures are allocated internally by uTask.
 */
typedef struct utask
{
	///task internal state (UTS_* defines) - internal data, do not modify manually
    volatile	utask_state_t	state;
	///internal task state (for programmers use)
    volatile	utask_state_t	istate;
	#ifdef UTASK_SEM
		///semaphore pointer - internal data, do not modify manually
		volatile 	utask_sem_t*	sem;
		#ifdef UTASK_SEM_CHANGE
			///samaphore conditional value - internal data, do not modify manually
			volatile utask_sem_t 	sem_val;
		#endif
	#endif
	///sleep counter (in ticks) - internal data, do not modify manually
    volatile	utask_timer_t	sleep;
	///function to be run
    void		(*fun)(struct utask* t);

    #ifdef UTASK_DATA
    	#if UTASK_DATA > 0
		///user's general purpose per task data storage
		utask_data_t	data[UTASK_DATA];
		#endif
	#endif

	#ifdef UTASK_ARG
		///user's data passed as param for the task
		utask_data_t	arg;
	#endif

	#ifdef UTASK_TID
		///task id
		utask_data_t	tid;
	#endif

	#ifdef UTASK_NAME
		///task name
        const char*		name;
    #endif
} utask_t;

///task function typedef
typedef void (*utask_fun)(utask_t* t);

/**
@brief Intializes internal variables. Must be run before all other utask functions.

Intializes internal variables. @b Must be run before all other utask functions.
@return nothing
*/
void	utask_init(void);


/**
@brief Scheduling function, must be called forever.

Scheduling function, @b must be called forever (e.g. in main() in while(1) {} loop) or utask will @b not work!
@return nothing
 */
void 	utask_schedule(void);


/**
@brief Timekeeping function, should be called from ISR.

Function decreasing tasks' sleep timers. @b Must be called periodically (e.g. from timer ISR) or utask_sleep() and utask_wait_*() will @b not work properly! 
Calling frequency selects timeout tick resolution. Calling this function at 1kHz rate gives you tick equal to 1 ms. Call this function as often (or as rarely) as you need.
@return nothing
*/
void	utask_sleep_process(void);


/**
@brief Add new task with function @a function for execution.

Add new task with function @a function for execution.
Can be one of two versions:
- simple version allowing adding tasks only before first call to utask_schedule(), enabled when @ref UTASK_SIMPLE_ADD is defined.
- version allowing spawning new task any time - when @ref UTASK_SIMPLE_ADD is undefined.

After add is called following structure members are set to as follows:
- state is set to UTS_READY and task will be executed as soon as possible,
- istate is set to zero,
- sleep is set to zero,
- arg (if enabled) is set to zero,
- tid (if enabled) is last task's id + 1,
- not mentioned members are in undefined state.

@param	fun			pointer to function to be executed
@return pointer to task structure describing task of NULL in case of error
*/
utask_t*	utask_add(utask_fun fun);

#ifdef UTASK_NAME
/**
@brief Add new task with function @a function and name @a name for execution.

Add new task with function @a function and name @a name for execution.
This is simple wrapper. See @ref utask_add() for details.
@param	fun			pointer to function to be executed
@param name		name of the task
@return pointer to task structure describing task of NULL in case of error
*/
utask_t*	utask_add_name(utask_fun fun, const char *name);
#endif


#ifdef UTASK_EXIT
/**
@brief Terminate task @a t.

Ends task @a t. utask_exit() can terminate other tasks. Do @b not reference @a t after it has been freed with utask_exit().
@param	t			pointer to utask structure
 */
void	utask_exit(utask_t* t);
#endif


/**
@brief Sleep task @a t for @a ticks ticks of clock.

Sleep task @a t for @a ticks ticks of clock. This function can put to sleep another task.
@param t pointer to utask structure
@param ticks number of ticks of clock for task to sleep
*/
void 	utask_sleep	(utask_t* t,utask_timer_t ticks);


#ifdef UTASK_SUSPEND_RESUME
/**
@brief Suspend task exectuion.

Suspend task exectuion. This function can suspend another task.
@param	t			pointer to utask structure
 */
void	utask_suspend(utask_t* t);


/**
@brief Resume task @a t.

Resume task @a t. It works on both on tasks put to sleep and suspended.
@param t	pointer to utask structure
*/
void utask_resume(utask_t* t);
#endif

/**
@brief Wait for the semaphore \a sem to be cleared with \a timeout ticks of timeout.

Wait for the semahore \a sem to be cleared with \a timeout ticks of timeout. If condition was met t->sleep != 0.
@param	t			pointer to utask structure
@param	sem			pointer to semaphore
@param	timeout	waiting timeout
*/
void	utask_wait_zero(utask_t* t, utask_sem_t* sem, utask_timer_t timeout);

/**
@brief Wait for the semahore \a sem to be set with \a timeout ticks of timeout.

Wait for the semahore \a sem to be set with \a timeout ticks of timeout. If condition was met t->sleep != 0.
@param	t			pointer to utask structure
@param	sem			pointer to semaphore
@param	timeout	waiting timeout
@return nothing
*/
void	utask_wait_nzero(utask_t* t,utask_sem_t* sem, utask_timer_t timeout);

#ifdef UTASK_SEM_CHANGE
/**
@brief Wait for the semaphore \a sem value to be set to \a sem_value with \a timeout ticks of timeout.

Wait for the semaphore \a sem value to be set to \a sem_value with \a timeout ticks of timeout. If contidion was met t->sleep != 0.
@param	t			pointer to utask structure
@param	sem			pointer to semaphore
@param	sem_val	value waited for
@param	timeout	waiting timeout
@return nothing
 */
void	utask_wait_eq(utask_t* t, utask_sem_t* sem, utask_sem_t sem_val, utask_timer_t timeout);

/**
@brief Wait for the semaphore \a sem value to be changed from \a sem_value with \a timeout ticks of timeout.

Wait for the semaphore \a sem value to be changed from \a sem_value with \a timeout ticks of timeout. If contidion was met t->sleep != 0.
@param	t			pointer to utask structure
@param	sem			pointer to semaphore
@param	sem_val 	value waited for
@param	timeout	waiting timeout
@return nothing
 */
void	utask_wait_neq(utask_t* t, utask_sem_t* sem, utask_sem_t sem_val, utask_timer_t timeout);
#endif


#ifdef UTASK_STATS
/**
@brief Returns free task slots.

Returns free task slots.
@return	free task slots
 */
utask_data_t	utask_get_free_slots(void);


/**
@brief Function returns number of tasks executing @a fun function.

Function returns number of tasks executing @a fun function.
@param   fun	function
@return	numbers of tasks executing @a fun function.
*/
utask_data_t	utask_get_task_cnt(utask_fun fun);
#endif

#ifdef UTASK_TID
	/**
	 @brief Returns task structure pointer by task id.
	 
	 Returns task structure pointer by task id.
	 @param	tid		task id
	 @return	pointer to the task structure or NULL if not found
	*/
	utask_t*	utask_get_by_tid(utask_data_t tid);
#endif


#endif					//UTASK_H


/**
@example utask_example_1.c
This simple example shows how to make a LED blink at 1 Hz rate using utasks.

@example utask_example_2.c
This example shows how to delay task execution by using utask_sleep().

@example utask_example_3.c
This is simple example of semaphor usage. See docs for more details.

@example utask_example_4.c
Example of spawning new tasks from running ones.
*/
